lib/ansible/modules/cloud/amazon/ec2_vpc_subnet_facts.py

```diff
@@ -77,73 +81,150 @@
+            returned: always
+            type: complex
+            contains:
+                association_id:
+                    description: The association ID
+                    returned: always
+                    type: string
+                ipv6_cidr_block:
+                    description: The IPv6 CIDR block that is associated with the subnet.
+                    returned: always
+                    type: string
+                ipv6_cidr_block_state:
+                    description: A hash/dict that contains a single item. The state of the cidr block association.
+                    returned: always
+                    type: dict
+                    contains:
+                        state:
+                            description: The CIDR block association state.
+                            returned: always
+                            type: string
+'''
 
+import traceback
 from ansible.module_utils.basic import AnsibleModule
-from ansible.module_utils.ec2 import AnsibleAWSError, connect_to_aws, ec2_argument_spec, get_aws_connection_info
-
+from ansible.module_utils.ec2 import (
+    boto3_conn,
+    ec2_argument_spec,
+    get_aws_connection_info,
+    AWSRetry,
+    HAS_BOTO3,
+    boto3_tag_list_to_ansible_dict,
+    camel_dict_to_snake_dict,
+    ansible_dict_to_boto3_filter_list
+)
 
-def get_subnet_info(subnet):
-
-    subnet_info = {'id': subnet.id,
-                   'availability_zone': subnet.availability_zone,
-                   'available_ip_address_count': subnet.available_ip_address_count,
-                   'cidr_block': subnet.cidr_block,
-                   'default_for_az': subnet.defaultForAz,
-                   'map_public_ip_on_launch': subnet.mapPublicIpOnLaunch,
-                   'state': subnet.state,
-                   'tags': subnet.tags,
-                   'vpc_id': subnet.vpc_id}
+try:
+    from botocore.exceptions import ClientError, NoCredentialsError
+except ImportError:
+    pass  # caught by imported HAS_BOTO3
 
-    return subnet_info
 
+@AWSRetry.backoff()
+def describe_subnets(connection, module):
+    """
+    Describe Subnets.
 
-def list_ec2_vpc_subnets(connection, module):
+    module  : AnsibleModule object
+    connection  : boto3 client connection object
+    """
+    # collect parameters
+    filters = ansible_dict_to_boto3_filter_list(module.params.get('filters'))
+    subnet_ids = module.params.get('subnet_ids')
 
-    filters = module.params.get("filters")
-    subnet_dict_array = []
+    # init empty list for return vars
+    subnet_info = list()
 
+    # Get the basic VPC info
     try:
-        all_subnets = connection.get_all_subnets(filters=filters)
-    except BotoServerError as e:
-        module.fail_json(msg=e.message)
+        response = connection.describe_subnets(SubnetIds=subnet_ids, Filters=filters)
+    except ClientError as e:
+        module.fail_json(msg=e.message, exception=traceback.format_exc(), **camel_dict_to_snake_dict(e.response))
```

Comment:
Please document how this should be done properly in the AWS guidelines doc, so that people (including me) don't have to reinvent the wheel every time.